File: java/src/main/java/demo/officeDay/linqVsStream/PersonStatistics.java
package demo.officeDay.linqVsStream;

import java.io.InputStream;
import java.util.*;
import java.util.stream.Collector;
import java.util.stream.Collectors;

public class PersonStatistics {
    private List<Person> personList = new ArrayList<>();

    public Map<String, Optional<Person>> oldestPersonByCountry(){
        return personList.stream().collect(Collectors.groupingBy(Person::getCountry, Collectors.maxBy(Comparator.comparingInt(Person::getAge))));
    }

    public Map<String, List<Person>> oldestPeopleByCountryImperative() {
        Map<String, List<Person>> peopleByCountry = new HashMap<>();
        for(Person person : personList){
            List<Person> people = peopleByCountry.get(person.getCountry());

            if(people == null) {
                people = new ArrayList<>();
                people.add(person);
                peopleByCountry.put(person.getCountry(), people);
                continue;
            }

            Person personFromList = people.get(0);
            if(personFromList.getAge() < person.getAge()){
                people.clear();
                people.add(person);
            } else if(personFromList.getAge() == person.getAge()){
                people.add(person);
            }
        }
        return peopleByCountry;
    }


    public Map<String, List<Person>> oldestPeopleByCountryCustomCollector(){
        return personList.stream().collect(Collectors.groupingBy(Person::getCountry, Collector.of(
                () -> new ArrayList<>(),
                (people, person) -> {
                    if(people.isEmpty()) {
                        people.add(person);
                        return;
                    }
                    Person personFromList = people.get(0);
                    if(personFromList.getAge() < person.getAge()){
                        people.clear();
                        people.add(person);
                    } else if(personFromList.getAge() == person.getAge()){
                        people.add(person);
                    }
                },
                (people, people2) -> {
                    people.addAll(people2);
                    return people;
                }
        )));
    }

    public Map<String, List<Person>> oldestPeopleByCountry(){
        return personList.stream()
                .collect(Collectors.collectingAndThen(Collectors.groupingBy(Person::getCountry),
                    (peopleByCountry) -> {
                        peopleByCountry.forEach((country, people)-> {
                            int maxAge = people.stream()
                                    .mapToInt(Person::getAge)
                                    .max().getAsInt();

                            List<Person> oldestPeople = people.stream()
                                    .filter((person) -> person.getAge() == maxAge)
                                    .collect(Collectors.toList());

                            peopleByCountry.put(country, oldestPeople);
                        });
                        return peopleByCountry;
                    }));
    }

    public PersonStatistics(InputStream source){
        Scanner scanner = new Scanner(source);
        while(scanner.hasNext()){
            String[] lineTokens = scanner.nextLine().split(",");
            personList.add(new Person(lineTokens[0], lineTokens[1], Integer.valueOf(lineTokens[2])));
        }
        scanner.close();
    }

}

File: csharp/demo.officeDay/linqVsStream/PersonStatistics.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace demo.officeDay.linqVsStream
{
    public class PersonStatistics
    {
        private List<Person> personList = new List<Person>();

        public PersonStatistics(string csvFile)
        {             
            personList.AddRange(
                File.ReadAllLines(csvFile)
                    .Select(line => new Person(line))
                    .ToList()); 
        }

        public Dictionary<String,Person> oldestPersonByCountry()
        {
            var oldestPersonByCountry =
                    from person in personList
                    group person by person.Country into countries
                    select new
                    {
                        Country = countries.Key,
                        Person = countries.First(personByCountry => personByCountry.Age == countries.Max(i => i.Age))
                    }
                ;

       
         /*   var oldestPersonByCountry =
            personList
                .GroupBy(person => person.Country)
                .Select(country => new
                {
                    Country = country.Key,
                    Person = country.First(i=>i.Age==country.Max(c=>c.Age))
                });
                */
            return oldestPersonByCountry
                   .ToDictionary(i => i.Country, i => i.Person);
        }

        public Dictionary<String, List<Person>> oldestPeopleByCountryImperative()
        {
            var peopleByCountry = new Dictionary<string, List<Person>>();
            foreach (var person in personList)
            {
                List<Person> people;

                if (!peopleByCountry.ContainsKey(person.Country))
                {
                    people = new List<Person>();
                    people.Add(person);
                    peopleByCountry.Add(person.Country, people);
                    continue;
                }
                else
                {
                    people = peopleByCountry[person.Country];
                }

                var personFromList = people[0];
                if (personFromList.Age < person.Age)
                {
                    people.Clear();
                    people.Add(person);
                }
                else if (personFromList.Age == person.Age)
                {
                    people.Add(person);
                }
            }
            return peopleByCountry;
        }

        public Dictionary<String, IEnumerable<Person>> oldestPeopleByCountry()
        {
            var oldestPeopleByCountry = 
                    from person in personList
                    group person by person.Country into countries
                    select new
                    {
                        Country = countries.Key,
                        Persons = countries.Where(personByCountry => personByCountry.Age==countries.Max(i=>i.Age))
                    }
            ;

            /*
            var oldestPeopleByCountry = personList
                .GroupBy(person => person.Country)
                .Select(country => new
                {
                    Country = country.Key,
                    Persons = country.Where(personByCountry => 
                                            personByCountry.Age == country.Max(i => i.Age))
                });
                */

            return oldestPeopleByCountry.ToDictionary(i=>i.Country,i=>i.Persons);
        }


    }
}
